frontend/javascripts/oxalis/model/sagas/isosurface_saga.js

```diff
@@ -369,6 +374,168 @@ function* maybeLoadIsosurface(
+    yield* call(ensureSuitableIsosurface, position, cellId, shouldBeRemoved);
+    shouldBeRemoved = false;
+  }
+  yield* put(finishedLoadingIsosurfaceAction(layerName, cellId));
+}
+
+/*
+ *
+ * Precomputed Meshes
+ *
+ */
+
+function* loadPrecomputedMesh(action: LoadPrecomputedMeshAction) {
+  const { cellId, seedPosition, meshFileName, layerName } = action;
+  const layer = yield* select(state =>
+    layerName != null
+      ? getSegmentationLayerByName(state.dataset, layerName)
+      : getVisibleSegmentationLayer(state),
+  );
+  if (layer == null) return;
+
+  // If a REMOVE_ISOSURFACE action is dispatched and consumed
+  // here before loadPrecomputedMeshForSegmentId is finished, the latter saga
+  // should be canceled automatically to avoid populating mesh data even though
+  // the mesh was removed. This is accomplished by redux-saga's race effect.
+  yield* race({
+    loadPrecomputedMeshForSegmentId: _call(
+      loadPrecomputedMeshForSegmentId,
+      cellId,
+      seedPosition,
+      meshFileName,
+      layer,
+    ),
+    cancel: _take(
+      otherAction =>
+        otherAction.type === "REMOVE_ISOSURFACE" &&
+        otherAction.cellId === cellId &&
+        otherAction.layerName === layerName,
+    ),
+  });
+}
+
+function* loadPrecomputedMeshForSegmentId(
+  id: number,
+  seedPosition: Vector3,
+  fileName: string,
+  segmentationLayer: APIDataLayer,
+): Saga<void> {
+  const layerName = segmentationLayer.name;
+  yield* put(addIsosurfaceAction(layerName, id, seedPosition, true));
+  yield* put(startedLoadingIsosurfaceAction(layerName, id));
+  const dataset = yield* select(state => state.dataset);
+
+  let availableChunks = null;
+  try {
+    availableChunks = yield* call(
+      getMeshfileChunksForSegment,
+      dataset.dataStore.url,
+      dataset,
+      getBaseSegmentationName(segmentationLayer),
+      fileName,
+      id,
+    );
+  } catch (exception) {
+    console.warn("Mesh chunk couldn't be loaded due to", exception);
+    Toast.warning(messages["tracing.mesh_listing_failed"]);
+
+    yield* put(finishedLoadingIsosurfaceAction(layerName, id));
+    yield* put(removeIsosurfaceAction(layerName, id));
+    return;
+  }
+
+  // Sort the chunks by distance to the seedPosition, so that the mesh loads from the inside out
+  const sortedAvailableChunks = _.sortBy(availableChunks, chunkPosition =>
+    V3.length(V3.sub(seedPosition, chunkPosition)),
+  );
+  const tasks = sortedAvailableChunks.map(chunkPosition => async () => {
+    const stlData = await getMeshfileChunkData(
+      dataset.dataStore.url,
+      dataset,
+      getBaseSegmentationName(segmentationLayer),
+      fileName,
+      id,
+      chunkPosition,
+    );
+
+    const geometry = parseStlBuffer(stlData);
+    getSceneController().addIsosurfaceFromGeometry(geometry, id);
+  });
+
+  try {
+    yield* call(processTaskWithPool, tasks, PARALLEL_PRECOMPUTED_MESH_LOADING_COUNT);
```

Comment:
Are you sure that this function is cancelled once it was kicked off? I would have assumed that once an element from `tasks` has been started, that it's not cancelable by the redux sagas (a symptom of this is that the tasks are `async`s and not generator functions). The proper way to fix this would probably be to convert `processTaskWithPool` to be saga-based. It's not used anywhere else, so adapting callers wouldn't be a problem. However, adapting the tests could get a bit annoying..

A simple alternative would obviously be to re-introduce some of the "is canceled"-checks :see_no_evil:

Reply:
Good point, I'll have a look and test what happens, but you're very likely correct.

> However, adapting the tests could get a bit annoying..

I don't think there are any tests for the mesh functionality yet, at least I haven't come across them :sweat_smile:
My plan is to add a screenshot test for this functionality. However, I'll also try to add a couple "higher-level" saga tests and will see whether it is too much "pain".

Reply:
> I don't think there are any tests for the mesh functionality yet, at least I haven't come across them sweat_smile

I meant the unit tests which exist for `processTaskWithPool` :)